Make sure collection is a list if a str is given

Shrews opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Supplying the collection name as a string instead of a list would cause a traceback
when we attempted to append the 'ansible.legacy' collection to the list we expected.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #69054

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
collections

##### ADDITIONAL INFORMATION

test.yaml
```
---
- hosts: localhost
  gather_facts: no
  collections: foo
  tasks:
    - debug: msg="Test"
```

results in:
```
  File "/Users/shrews/Devel/github/Shrews/ansible/lib/ansible/playbook/collectionsearch.py", line 31, in _ensure_default_collection
    collection_list.append('ansible.legacy')
AttributeError: 'AnsibleUnicode' object has no attribute 'append'
```

Shrews committed:
Make sure collection is a list if a str is given

sivel commented on lib/ansible/playbook/collectionsearch.py:
```diff
@@ -19,12 +19,14 @@ def _ensure_default_collection(collection_list=None):
-            collection_list = [collection_list]
+    # We allow collections to be either a string or a list, so make sure
+    # we have a list going forward.
+    if isinstance(collection_list, string_types):
```
Should we insert an `elif not isinstance(collection_list, (None, list)):` and fail?  As it's written, someone could supply all sorts of other incorrect data types.

Shrews replied:
Good idea. If I make it a dict, we get the same traceback about not being able to call append(). I'll make that change.

bcoca replied:
instead of custom code, why just not force validation? this would handle 'non lists' already correctly and since 'collections' is static, we don't  have to worry about templating

Shrews replied:
@bcoca If we assume CollectionSearch is always mixed in with Base (seems like it is?), then we can call self.get_validated_value() early on. I'm unclear on if the mixin is a safe assumption going forward though.

bcoca replied:
i really want to redesign it so 'get_validated_value' is the normal 'getter' and that 'get_unvalidated' becomes the method for when we need 'direct' access.

Shrews committed:
Call field validation early on collections

Because we are doing work on modifying the collections value before
it is actually validated, we can validate it ourselves early to make
sure the user supplies either a string or list. Dicts are not valid.

The new validation allows us to simplify the _ensure_default_collection()
function. And since the field is now static, we no longer need to specify
a default for it, which also allows us to simplify the function. Since
the default is now removed, we can also remove the sanity/ignore.txt entry
for collectionsearch.py.

New unit tests are added (and the existing one modified) that allow us to
make sure that we throw a parser error if a user specifies something other
than a string or list for the collections value everywhere it can be specified.

Shrews committed:
Revert removing the collection default

The default is actually used, so restore it.

Shrews committed:
Fix unit tests in test_helpers.py affected by early collection validation
